### ISSUE TITLE
Non-deterministic error using batch receiver

### ISSUE THREAD

**Issue by pstrato-agrimetrics** [Azure/azure-event-hubs-python#42]
Based on examples/recv_batch.py, I'm trying to retrieve all events with the following code:
```
from datetime import datetime
from azure.eventhub import EventHubClient, Offset

CONSUMER_GROUP = "$Default"
PARTITION = "0"

client = EventHubClient(ADDRESS, debug=True, username=USER, password=KEY)

try:
    receiver = client.add_receiver(CONSUMER_GROUP, PARTITION, offset=Offset(datetime(2018, 7, 9)))
    client.run()
    total = 0
    end = False
    while not end:
        events = receiver.receive(max_batch_size=20, timeout=1)
        if len(events) > 0:
            for event_data in events:
                event_offset = event_data.offset
                event_message = str(event_data.message)
                total = total + 1

        else:
            end = True
        print('received', total)
except KeyboardInterrupt as e:
    end = True
finally:
    client.stop()

print('found ', total, 'report')
```

The program above sometimes complete (processing more than 80,000 rows), sometimes doesn't with the following error:

Windows:
```
Process finished with exit code -1073740940 (0xC0000374)
```
We are using azure.eventhub 0.2.0rc1 with python 3.6.5 on Windows 10 Pro.
Please let me know of any additional information you would like to have.

**Comment by annatisch** [comment]
Thanks for the issue report @pstrato-agrimetrics!
A couple of question:
- Was this running successfully with an earlier release? Or is RC2 the first release you've tried?
- I see you've set debug to True - are you running with a logger? Would it be possible to see these logs?
- Does the error seem to occur at a regular point (e.g. after 5 seconds) or does it appear entirely random?

Thanks!

**Comment by pstrato-agrimetrics** [comment]
Using 0.2.0b2, I get the same error, e.g.:
```
…
received 667
Process finished with exit code -1073740940 (0xC0000374)
```

With rc1, memory usage remains constant, approx. 18.9 MB.

With both versions (b2 and RC1 (I cannot see a RC2 available)), the error appears at random time/point, sometimes the number of messages retrieved is small (~600) sometimes bigger (~8000).
 Often, the program terminates without error.
I attach the logs for both b2 and RC1.

[0.2.0b2.zip](https://github.com/Azure/azure-event-hubs-python/files/2183737/0.2.0b2.zip)
[0.2.0rc1.zip](https://github.com/Azure/azure-event-hubs-python/files/2184046/0.2.0rc1.zip)

**Comment by afernando79** [comment]
The issue seems to be more apparent in Linux / iOS. 

I have used the following with success albeit with poor performance:

```
ADDRESS = os.environ.get('EVENT_HUB_ADDRESS')
USER = os.environ.get('EVENT_HUB_SAS_POLICY')
KEY = os.environ.get('EVENT_HUB_SAS_KEY')
CONSUMER_GROUP = "$default"
OFFSET = Offset("-1")
PARTITION = "0"

total = 0
last_sn = -1
last_offset = "-1"
client = EventHubClient(ADDRESS, debug=False, username=USER, password=KEY)
try:
    receiver = client.add_receiver(CONSUMER_GROUP, PARTITION, prefetch=1, offset=OFFSET)
    client.run()
    while(1):
        batched_events = receiver.receive(max_batch_size=None, timeout=10)
        for event_data in batched_events:
            last_offset = event_data.offset
            last_sn = event_data.sequence_number
            total += 1
            print("Partition {}, Received {}, sn={} offset={}".format(
                PARTITION,
                total,
                last_sn,
                last_offset))

except KeyboardInterrupt:
    pass
finally:
    client.stop()
```

Is there an issue with the way in which the messages / eventhub objects are being placed on the stack?

In Windows, I am using prefetch=100 without any obvious issues.

For both Windows and Linux I am using the RC1.

**Comment by afernando79** [comment]
Yes, prefetch > 1 in Linux fails after an indeterminate amount of time. Could be 5 seconds or 50. The issue appeared to worsen as prefetch was increased.

On 11 Jul 2018 17:25, annatisch <notifications@github.com> wrote:

Thanks for the input @afernando79<https://github.com/afernando79>! Just to confirm I'm following correctly - your observation is that the larger the prefetch value, the higher the likelyhood of crashing? And with a prefetch of 1, no errors were encountered?

**Comment by nimonika** [comment]
On macOS When running the above code, I keep getting 
```
python(6241,0x10189c340) malloc: *** error for object 0x108366e38: incorrect checksum for freed object - object was probably modified after being freed.
*** set a breakpoint in malloc_error_break to debug
```
There certainly seems to be an issue with RC1 as this error was not showing up in the 0.2.0b2.

**Comment by afernando79** [comment]
As a followup, although the above code works, it will fail even with prefetch=1 if "event_data.message" is accessed.

Reverting to 0.2.0b2 (and removing incompatible function calls, eg offset) resolves the issue.

Hope this helps focus your efforts.

**Comment by nimonika** [comment]
On  Ubuntu 16.04, a Segmentation fault (core dumped) shows up even for 0.2.0b2.

**Comment by annatisch** [comment]
Thanks for the additional feedback @afernando79, @nimonika!
So it appears to be related to the EventData object - and retrieving various attributes.
You shouldn't need to access `EventData.message` directly - unless there's a particular attribute not exposed that you need?

I haven't been able to reproduce the crashing yet - still working on it :)

**Comment by afernando79** [comment]
Thanks for continuing to look into this @annatisch. Forgive the ignorance but how can the message content be accessed if not directly? Can you please suggest a code snippet?

Also, what OS are you running the above code in?

**Comment by annatisch** [comment]
Thanks @afernando79 - try using `EventData.body` - depending on how you created the events/messages this will contain either `bytes` or a `generator` (in most EH scenarios this will be bytes I think).

I do most of my development on Windows and OSX - but run some of the scenario tests on all three of Windows, OSX and Ubuntu before publishing a release. Though clearly there's either a scenario that my tests don't cover, or I don't run them for long enough to hit this issue.... :(

**Comment by annatisch** [comment]
So this sample seems to run indefinitely on my test machines - both Windows and Linux.

I'm wondering if each of @afernando79, @pstrato-agrimetrics and @nimonika could give me a sample of how you're _creating and sending_ messages to the eventhub.

Perhaps this error is related to the message content and how the receive client is attempting to deserialize it.

**Comment by nimonika** [comment]
@annatisch The messages are JSON objects which send to the EH through a policy defined in the API management layer. How does one get access to the generator object from EventData.body? Could you pass us some example code? Thanks!

**Comment by nimonika** [comment]
@annatisch Thank you for your advise about using generator. That works, however I still cannot get rid of the malloc error.

**Comment by annatisch** [comment]
Thanks @nimonika!
Do you know approximately how large the json objects is each message are?

**Comment by pstrato-agrimetrics** [comment]
Our json objects are approximately 540 characters long. I confirm that using the generator code does not fix the issue.

**Comment by annatisch** [comment]
Thank you all for your feedback and patience!
I believe I have now reproduced the error! Debugging now :)

**Comment by annatisch** [comment]
Okay - I have found and fixed the issue.
I have a couple of other commits to make and will try to publish a new release on Monday :)

For anyone wanting to patch a fix in the meantime to confirm their issue is the same one I found, you can open up `message.py` in the `uamqp` library in your pip packages environment and comment out these two lines:
https://github.com/Azure/azure-uamqp-python/blob/master/uamqp/message.py#L147-L148

```
if _app_props:
    self.application_properties = _app_props.map
```

**Comment by andyamos79** [comment]
Great news @annatisch! Great support.

I'll check on Monday.

Thanks again!

**Comment by nimonika** [comment]
@annatisch You seemed to have nailed it!Works without any errors on mine now!
